// coinbase-pro-node/dist/demo.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const CoinbasePro_1 = require("./CoinbasePro");
const product_1 = require("./product");
require('dotenv').config();
let client;
if (process.env.USE_SANDBOX === 'true') {
    console.info("Using Coinbase Pro's public sandbox with API key...");
    client = new CoinbasePro_1.CoinbasePro({
        apiKey: process.env.COINBASE_PRO_API_KEY,
        apiSecret: process.env.COINBASE_PRO_API_SECRET,
        passphrase: process.env.COINBASE_PRO_PASSPHRASE,
        useSandbox: true,
    });
}
else if (process.env.USE_SANDBOX === 'false') {
    console.info("Using Coinbase Pro's production environment with API key...");
    client = new CoinbasePro_1.CoinbasePro({
        apiKey: process.env.COINBASE_PRO_SANDBOX_API_KEY,
        apiSecret: process.env.COINBASE_PRO_SANDBOX_API_SECRET,
        passphrase: process.env.COINBASE_PRO_SANDBOX_PASSPHRASE,
        useSandbox: false,
    });
}
else {
    console.info("Using Coinbase Pro's production environment without API key...");
    client = new CoinbasePro_1.CoinbasePro();
}
function main() {
    return __awaiter(this, void 0, void 0, function* () {
        const candles = yield client.rest.product.getCandles('BTC-USD', {
            end: '2020-04-11T10:00:00.000Z',
            granularity: product_1.CandleGranularity.ONE_HOUR,
            start: '2020-04-11T08:00:00.000Z',
        });
        console.info(`Received "${candles.length}" candles to represent 3 hours (08 - 11 AM).`, candles);
    });
}
main().catch(console.error);
